haystack/preview/components/builders/dynamic_prompt_builder.py

```diff
@@ -0,0 +1,334 @@
+        component.set_input_types(self, **run_input_slots, **kwargs_input_slots)
+
+        # setup outputs
+        if chat_mode:
+            component.set_output_types(self, prompt=List[ChatMessage])
+        else:
+            component.set_output_types(self, prompt=str)
+
+        self.expected_runtime_variables = expected_runtime_variables
+        self.chat_mode = chat_mode
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+         Converts the `DynamicPromptBuilder` instance to a dictionary format, primarily for serialization purposes.
+
+        :return: A dictionary representation of the `DynamicPromptBuilder` instance, including its template variables.
+        :rtype: Dict[str, Any]
+        """
+        return default_to_dict(
+            self, expected_runtime_variables=self.expected_runtime_variables, chat_mode=self.chat_mode
+        )
+
+    def run(
+        self,
+        prompt_source: Union[List[ChatMessage], str],
+        template_variables: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ):
+        """
+        Executes the dynamic prompt building process. Depending on the provided type of `prompt_source`, this method
+        either processes a list of `ChatMessage` instances or a string template. In the case of `ChatMessage` instances,
+        the last user message is treated as a template and rendered with the resolved pipeline variables and any
+        additional template variables provided. For a string template, it directly applies the template variables to
+        render the final prompt. You can provide additional template variables directly to this method, that are then merged
+        with the variables resolved from the pipeline runtime.
+
+        :param prompt_source: A list of `ChatMessage` instances or a string template. The list scenario assumes the last
+        user message as the template for the chat prompt, while the string scenario is used for non-chat-based prompts.
+        :type prompt_source: Union[List[ChatMessage], str]
+
+        :param template_variables: An optional dictionary of template variables. Template variables provided at
+        initialization are required to resolve pipeline variables, and these are additional variables users can
+        provide directly to this method.
+        :type template_variables: Optional[Dict[str, Any]]
+
+        :param kwargs: Additional keyword arguments, typically resolved from a pipeline, which are merged with the
+        provided template variables.
+
+        :return: A dictionary containing the key "prompt", which holds either the updated list of `ChatMessage`
+        instances or the rendered string template, forming the complete dynamic prompt.
+        :rtype: Dict[str, Union[List[ChatMessage], str]]
+        """
+        template_variables_combined = (
+            {**(kwargs or {}), **(template_variables or {})} if (template_variables or kwargs) else {}
+        )
```

Comment:
I would simplify this a bit, it's really hard to read this way. Something like this should work.

```suggestion
        kwargs = kwargs or {}
        template_variables = template_variables or {}
        template_variables_combined = {**kwargs, **template_variables}
```

Reply:
Ok true, I agree